File: CodeForces/Graph.cpp
#include <iostream>
#include <cstdlib>
#include <vector>
#include <algorithm>

using namespace std;

#define WHITE 0
#define GRAY 1
#define BLACK 2

class Node{
public:
  int n, color;
  NODE(){n=0; color = 0; }
  ~NODE(){}
};

class Graph{
private:
  long long int numerator, denominator;
  int V, E;
  vector<int> *G;
  Node *v;
  int time;
    
public:
  Graph();
  ~Graph();
  void dfs();
  void dfsVisit(int u);
};

void Graph::dfsVisit(int u){
  time++;
  v[u].color = GRAY;

  vector<int>::iterator i;
  for(i=G[u].begin(); i!=G[u].end(); i++){
    int n = i->n;
    if(v[n].color == WHITE){  dfsVisit(n);}
  }
  v[u].color = BLACK;
  time++;
}

void Graph::dfs(){
  time = 0;

  for(int u=1; u<=V; u++){
    if(v[u].color == WHITE){
      dfsVisit(u);
    }
  }
}

Graph::Graph(){
  numerator = 0;
  denominator = 02;
  cin>>V;
  E = V-1;
  v = new Node[V+1];
  G = new vector<int>[V+1];
  int x, y;
  for(int i=0; i<E; i++ ){
    cin>>x>>y;
    G[x].push_back(y);
    G[y].push_back(x);
  }

  dfs();
}

Graph::~Graph(){
  if(v!=NULL){ delete[] v;}
  if(G!=NULL){ delete[] G;}
}

int main()
{
  Graph G;
  return 0;
}

File: 1-JavaCodes/Algorithms-in-Java/src/GraphGFSPrototype1/Graph.java
package GraphGFSPrototype1;

import java.util.Iterator;
import java.util.LinkedList;

/**
 * Created by Fahim on 11/14/2016.
 */
public class Graph {

    private int V;
    private LinkedList<Integer> adj[];
    private boolean unDir;
    private static int time;
    enum Color {white, gray, black};
    private Color color[];

    //constructor
    public Graph(int v, boolean b){
        time = 0;
        unDir = b;
        V = v;
        adj = new LinkedList[v];
        color = new Color[v];
        for(int i=0;i<v;i++)  {

            adj[i] = new LinkedList();
            //color[i];
        }
    }
    public void AddEdge(int v, int w){

        adj[v].add(w);
        if(unDir==true)
            adj[w].add(v);
    }

    void DFSUtil(int v, boolean visited[]){

        visited[v] = true;          //mark the current node as visited
        System.out.println(v);      // and print it
        time = time+1;
        int dt = time;
        System.out.println(v+"-->dt = "+dt);
        Iterator<Integer> i = adj[v].listIterator();
        while(i.hasNext()){

            int n = i.next();
            if(!visited[n]){

                DFSUtil(n, visited);
            }
        }
        time = time+1;
        int ft = time;
        System.out.println(v+"-->ft = "+ft);
    }

    public void DFS(){

        boolean visited[] = new boolean[V];
        //char color[] = new char[V];
        //for(int i=0; i<V; i++)


        for(int i=0;i<V; i++){

            if(visited[i] == false)
                DFSUtil(i, visited);
        }
    }
}


